### DISCUSSION TITLE
How to make babel plugins recognized by browserify?

### DISCUSSION THREAD

**Discussion by gogo9th** [babel/babel#12561] [Q&A] [answered]
Hi,

I am using two Babel plugins in the following manner:

```
                var res = babel.transform(code, {
                    retainLines: true,
                    compact: false,
                    sourceType: 'unambiguous',
                    sourceMaps: true,
                    // presets: ['@babel/preset-env'],
                    plugins: [
                        "@babel/transform-runtime",
                        ["transform-es2015-modules-simple-commonjs",
                            {
                                'noMangle': true,
                                'allowTopLevelThis': true,
                                'strictMode': false
                            },
                        ]
                    ]
                }).code;
```

However, I need to convert my NodeJS code with `browserify` to make it work on browsers, and if I browserify the above code, two plugins (`@babel/transform-runtime, transform-es2015-modules-simple-commonjs`) are not recognized by `browserify`. How can I make them compatible with `browserify`? For example, is there a way to include these plugins with `require()`?

**Comment by gogo9th** [comment, marked as answer]
I solved the problem by using [babelify](https://github.com/babel/babelify). But babelify seems to have a problem: after running babelify with presets or plugins, the global `this` in the original code points to a wrong value. I avoid this problem by replacing `this` with `globalThis` in the original code. The code works now, although I'm not sure if this is the right solution.

**Reply by nicolo-ribaudo** [reply]
The `allowTopLevelThis` option of `transform-es2015-modules-simple-commonjs` prevents it from transforming `this` to `undefined`, but then it's browserify that decides how to handle it.

Sicne browserify emulates CommonJS, I would expect `this` to be an alias to `module.exports` as it is in Node.js. If you want a reference to the global `this` value, then `globalThis` is the correct solution!

(I'm marking this as answered since "use babelify" is the answer that other people finding this discussion might need)
